File: Leetcode/19.删除链表的倒数第-n-个结点.go
/*
 * @lc app=leetcode.cn id=19 lang=golang
 *
 * [19] 删除链表的倒数第 N 个结点
 */
package leetcode
/* --- 2022-04-07-10-14 --- */
// https://github.com/fengwei2002/Algorithm

type ListNode struct {
	Val int
	Next *ListNode
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{
		Val: -1,
		Next: head,
	}
	slow := dummy
	fast := dummy
	for i := 0; fast.Next != nil && i < n; i++ {
		fast = fast.Next
	}
    
    for slow.Next != nil && fast.Next != nil {
        slow = slow.Next 
        fast = fast.Next 
    }
    
    slow.Next = slow.Next.Next

	return dummy.Next
}
// @lc code=end

// 给你一个链表，删除链表的倒数第 n 个结点，并且返回链表的头结点。

// 进阶：你能尝试使用一趟扫描实现吗？

 

File: Leetcode/19.删除链表的倒数第-n-个结点.cpp
/*
 * @lc app=leetcode.cn id=19 lang=cpp
 *
 * [19] 删除链表的倒数第 N 个结点
 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

/*2021-10-02-00-59*/

// from https://github.com/fengwei2002/Algorithm
/*
 !
 !
 * https://leetcode.cn/problems/remove-nth-node-from-end-of-list/solution/lc-19-fengwei2002-by-fengwei2002-e11l/
 *
 */
struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
   public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* extHead = new ListNode(-1);
        extHead->next = head;

        ListNode* first = extHead;
        ListNode* second = extHead;

        for (int i = 0; i <= n; i++) {
            first = first->next;
        }

        while (first != nullptr) {
            first = first->next;
            second = second->next;
        }

        second->next = second->next->next;

        return extHead->next;
    }
};
// @lc code=end
